pkg/auth/auth.go

```diff
@@ -147,91 +147,118 @@ var GetPermissions = func(visas Visas) []string {
-		visaClaim := verifiedVisa.PrivateClaims()["ga4gh_visa_v1"]
-		visa := Visa{}
-		visaClaimJSON, err := json.Marshal(visaClaim)
-		if err != nil {
-			log.Errorf("failed to parse visa claim to JSON, %s, %s", err, visaClaim)
-			continue
-		}
-		err = json.Unmarshal(visaClaimJSON, &visa)
-		if err != nil {
-			log.Errorf("failed to parse visa claim JSON into struct, %s, %s", err, visaClaimJSON)
-			continue
-		}
-		exists, err := database.CheckDataset(visa.Dataset)
-		if err != nil {
-			log.Debugf("visa contained dataset %s which doesn't exist in this instance, skip", visa.Dataset)
-			continue
-		}
-		if exists {
-			log.Debugf("checking dataset list for duplicates of %s", visa.Dataset)
-			// check that dataset name doesn't already exist in return list,
-			// we can get duplicates when using multiple AAIs
-			duplicate := false
-			for i := range datasets {
-				if datasets[i] == visa.Dataset {
-					duplicate = true
-					log.Debugf("found a duplicate: dataset %s is already found, skip", visa.Dataset)
-					continue
-				}
-			}
-			if !duplicate {
-				log.Debugf("no duplicates of dataset: %s, add dataset to list of permissions", visa.Dataset)
-				datasets = append(datasets, visa.Dataset)
+	}
+
+	log.Debugf("matched datasets: %s", datasets)
+	return datasets
+}
+
+func checkVisaType(visa string, visaType string) bool {
+
+	log.Debug("checking visa type")
+
+	unknownToken, err := jwt.Parse([]byte(visa))
+	if err != nil {
+		log.Errorf("failed to parse visa, %s", err)
+		return false
+	}
+	unknownTokenVisaClaim := unknownToken.PrivateClaims()["ga4gh_visa_v1"]
+	unknownTokenVisa := Visa{}
+	unknownTokenVisaClaimJSON, err := json.Marshal(unknownTokenVisaClaim)
+	if err != nil {
+		log.Errorf("failed to parse unknown visa claim: %s to JSON, with error: %s", unknownTokenVisaClaim, err)
+		return false
+	}
+	err = json.Unmarshal(unknownTokenVisaClaimJSON, &unknownTokenVisa)
+	if err != nil {
+		log.Errorf("failed to parse unknown visa claim: %s to JSON, with error: %s", unknownTokenVisaClaim, err)
+		return false
+	}
+	if unknownTokenVisa.Type != visaType {
+		log.Debugf("visa is not of type: %s, skip", visaType)
+		return false
+	}
+	log.Debug("visa type check passed")
+
+	return true
+
+}
+
+func validateVisa(visa string) (jwt.Token, bool) {
+	log.Debug("start visa validation")
+	// Extract header from header.payload.signature
+	header, err := jws.Parse([]byte(visa))
+	if err != nil {
+		log.Errorf("failed to parse visa header, %s", err)
+		return nil, false
+	}
+	// Parse the jku key from header
+	o := OIDCDetails{
+		JWK: header.Signatures()[0].ProtectedHeaders().JWKSetURL(),
+	}
+	// Verify visa signature
+	verifiedVisa, err := VerifyJWT(o, visa)
+	if err != nil {
+		log.Errorf("failed to validate visa, %s", err)
+		return nil, false
+	}
+
+	if !ValidateTrustedIss(verifiedVisa.Issuer(), o.JWK) {
+		log.Infof("combination of iss: %s and jku: %s is not trusted", verifiedVisa.Issuer(), o.JWK)
+		return nil, false
+	}
```

Comment:
```suggestion
	if !ValidateTrustedIss(verifiedVisa.Issuer(), o.JWK) {
		log.Infof("combination of iss: %s and jku: %s is not trusted", verifiedVisa.Issuer(), o.JWK)
		return nil, false
	}

	// Verify visa signature
	verifiedVisa, err := VerifyJWT(o, visa)
	if err != nil {
		log.Errorf("failed to validate visa, %s", err)
		return nil, false
	}
```
The visa spec suggests, that we should check the trusted issuer before requesting JWK from JKU, because the URL can be malicious.